jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -267,46 +255,39 @@ private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> ent
 
         RetainableByteBuffer buffer = entry.getPooled();
         BufferUtil.reset(buffer.getByteBuffer());
-        boolean direct = buffer.isDirect();
-        int capacity = bucket.getCapacity();
-        long excessMemory = addMemoryAndGetExcess(bucket, direct);
-        if (excessMemory > 0)
-        {
-            bucket.recordEvict();
-            // If we cannot free enough space for the entry, remove it.
-            if (!evict(excessMemory, bucket, direct))
-            {
-                subtractMemory(capacity, direct);
-                bucket.recordRemove();
-                entry.remove();
-                return;
-            }
-        }
 
         // We have enough space for this entry, pool it.
         if (entry.release())
+        {
+            if (buffer instanceof Buffer b && b.use() % 100 == 1)
+               checkMaxMemory(buffer.isDirect());
             return;
+        }
 
         // Not enough space, discard this buffer.
-        subtractMemory(capacity, direct);
         bucket.recordRemove();
         entry.remove();
     }
 
-    private long addMemoryAndGetExcess(RetainedBucket bucket, boolean direct)
+    private void checkMaxMemory(boolean direct)
     {
-        long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
-        if (maxMemory < 0)
-            return -1;
-
-        AtomicLong memory = direct ? _directMemory : _heapMemory;
-        int capacity = bucket.getCapacity();
-        long newMemory = memory.addAndGet(capacity);
-        // Account the excess at most for the bucket capacity.
-        return Math.min(capacity, newMemory - maxMemory);
+        long max = direct ? _maxDirectMemory : _maxHeapMemory;
+        if (max <= 0 || !_evictor.compareAndSet(false, true))
+            return;
+        try
+        {
+            long memory = getMemory(direct);
+            long excess = memory - max;
+            if (excess > 0)
+                evict(excess, direct);
```

Comment:
I assume you removed the logic about `evict()` returning false to remove the current entry because the odds are too great that this would happen despite the eviction?

Reply:
Because we have already added the released buffer when we check for max memory, so we can't avoid churn.... but at least we would only churn on 1% and then only 1 thread.